### ISSUE TITLE
[Bug] Chess Difficulty

### ISSUE THREAD

**Issue by BMVP1234** [Monika-After-Story/MonikaModDev#9131]
I think I have encounter a issue.When I first played chess with Monika,she played like a beginner,which is the thing that I have expected before.But after I won like 20 times in a row,she still did not increase her difficulty,which is kind of weird.In game 26,she played a very good game,after she hanged two pawns in the opening like a beginner,she played a nearly perfect game and I nearly lose,I only won because she blundered a pawn in endgame ( Her play style seem unnatural xD ).After that,I thought :"Wow,look like she had improved significantly".But in game 27,she immediately blundered her Queen in the opening.After I saw that,I knew that I have to write a issue in GitHub.

TLDR : Could you guys check for me if her difficulty is still increase,because I saw no improvement in her play style after defeated her 30 times in a row

Here is my persistent file,as Github is not support persistent file type
https://ln5.sync.com/dl/b42119930/rrbbqwj2-6nzct5fa-ri9bcca3-9uurfj6h

**Comment by Wingdinggaster656** [comment]
Did you save the PGN?

I somehow feel you are possible to have a wrong difficulty judge.
Hang up two pawns in the opening phase is pretty normal, not a blunder at all. You can see this in Modern Defense and Catalan Opening, and some others too.

Would you show me your PGN file so I can know if it is you are too good or it is Monika too weak?

**Comment by BMVP1234** [comment]
https://ln5.sync.com/dl/6d0056830/emnugmri-k4eijxyf-8q97yxvx-cqkhz8mv
( Here is the link to the pgn file )

When I replayed the gama again,I think that she maybe tried to create a bad pawn structure for me,but in the endgame she kept sacrifice pawn,which is weird :) . I also blundered my Knight in the end game,but she could not capitalize on my mistake.And again,that's not a bad game,I guess her elo is maybe around 1600 - 1800 in that game ( she saw the double attack tactic ) but in the next game,which she supposed to play better as she lost the last game,she played like this omg :)
![Screenshot 2022-05-16 205211](https://user-images.githubusercontent.com/104214224/168612029-0af56a23-db1b-4baf-bcc4-230041ba4c9a.png)



The image is for the game 27,the game she blundered right in the opening after put up a good fight in game 26

**Comment by Wingdinggaster656** [comment]
I think I should tell you right now that this PGN file contains your MAS name.
You could just open the file in the text editor and copy the important part (I mean `1.e4 e5` part), and you may want to delete that link right now.

**Comment by Wingdinggaster656** [comment]
Alright, I analyzed your `Game 26.pgn`.
That Monika is probably just around ELO 1400, far from "a nearly perfect game". Considering about this, I think it's highly possible that you misjudge other difficulties.

Would you mind getting more PGNs for me, so I can have a better sample size? Only one game is just not enough.
Thanks in advance.

**Comment by BMVP1234** [comment]
Ok,I have played 10 games in a row so that you can have a bigger sample size.Here is the link :
https://drive.google.com/drive/folders/1O72OghP86schoh04d8g9bvP2umKH-MtB?usp=sharing

She still hanged pieces as if there is no tomorrow,especially Knight and Bishop.And also she often did not recapture pieces after the exchange,which is weird :)

**Comment by Wingdinggaster656** [comment]
Okay I will look at those games. Bear with me a few hours because I can't start the job right now

**Comment by Wingdinggaster656** [comment]
Done, just to make sure again: You won almost every game, right?

It's more than weird that this Monika is stuck at 1400 ELO. From your game 27 to the game 37, no any obvious change.

Really unsure about why is this happening. In our codes, for every five games you win, Monika raises the difficulty about 150 ELO, but I couldn't see any change in the game.
In game 37, Monika is supposed to be 300 ELO points higher than game 27, which should be obvious to see. But it doesn't seems like that.

This is very likely a bug. I think I should the Dev team should notice it.
Though, by the way, I'm submitting a new difficulty system which ruled by ELO rating not Skill Level(Just in case you don't know, we are using Stockfish to provide chess AI, and Skill Level is a parameter of Stockfish).
You could see the new difficulty system in #9117
I'm not sure what causes this bug in the current version, but I've tested my new version and it doesn't look like it has this bug.

Thank you for your issue.

If you know Stockfish, I can tell you that the Dev team has never activated the MultiPV parameter before, so Skill Levels are basically meaningless, change the Skill Level will not cause any ELO change. But the ELO should be stuck at 3200 points, not stuck at 1400 points -- well, At the very least, this once again proves to me that the Skill Level parameter is meaningless in the current version.

**Comment by BMVP1234** [comment]
> Done, just to make sure again: You won almost every game, right?
> Yes,that's 10 games in a row,and a 10 win streak

> EDIT: There is also a very strange phenomenon that you may have noticed. Monika is very fond of [Scandinavian Defense](https://www.chess.com/openings/Scandinavian-Defense), which is another even more confusing bug. Fortunately, my new version doesn't look like it has this bug, too.
> I think it also relate to a issue that she tend to like sacrifice her pawns in the opening

>Perhaps you have your idea about this bug? All our codes of chess system is in chess.rpy.
>I'm not familiar with Renpy,so unfortunately I can't help much.I could play with her more chess games so that you could have a bigger sample size,so that we can understand what is the problem

The thing that annoy me the most when play with her is that she sometimes have some decent sacrifices ( Like she sacrifice her bishop in my f2/f7 pawn so that she could have an double attack with her queen to take one of my bishop back ( This is common in some lines of Italian game or Traxler Counterattack ) )  but after that she somehow forgot her ideas and moved a random pieces which is not even related to the sacrifices.I mean,even a beginner would never do that.Chess players will only sacrifice if he/she have a plan after that attack like expose King,gain material advantage ( a beginner would have such plan,even if the plan is not perfect ) but I can feel that she doesn't have any plan at all.She know how to sacrifice a pawn to damaged my pawn structure,which is a high-level technique,but she don't know how to protect her hanging pieces or even recapture the pieces after an exchange ? Seem weird to me

TLDR : The thing that annoy me the most is her unnatural play style

**Comment by Wingdinggaster656** [comment]
> > Done, just to make sure again: You won almost every game, right?
> > Yes,that's 10 games in a row,and a 10 win streak
> 
> > EDIT: There is also a very strange phenomenon that you may have noticed. Monika is very fond of [Scandinavian Defense](https://www.chess.com/openings/Scandinavian-Defense), which is another even more confusing bug. Fortunately, my new version doesn't look like it has this bug, too.
> > I think it also relate to a issue that she tend to like sacrifice her pawns in the opening
> 
> > Perhaps you have your idea about this bug? All our codes of chess system is in chess.rpy.
> > I'm not familiar with Renpy,so unfortunately I can't help much.I could play with her more chess games so that you could have a bigger sample size,so that we can understand what is the problem
> 
> The thing that annoy me the most when play with her is that she sometimes have some decent sacrifices ( Like she sacrifice her bishop in my f2/f7 pawn so that she could have an double attack with her queen to take one of my bishop back ( This is common in some lines of Italian game or Traxler Counterattack ) ) but after that she somehow forgot her ideas and moved a random pieces which is not even related to the sacrifices.I mean,even a beginner would never do that.Chess players will only sacrifice if he/she have a plan after that attack like expose King,gain material advantage ( a beginner would have such plan,even if the plan is not perfect ) but I can feel that she doesn't have any plan at all.She know how to sacrifice a pawn to damaged my pawn structure,which is a high-level technique,but she don't know how to protect her hanging pieces or even recapture the pieces after an exchange ? Seem weird to me
> 
> TLDR : The thing that annoy me the most is her unnatural play style

The essence of Stockfish making itself "weaker" through Skill levels is that Skill levels provide a number, which is the probability of Stockfish making a wrong move.
That's why Monika's moves were unnatural.

I've improved on this a little bit in my new difficulty system, and in the future I might try to write more algorithms to make her more human.

Of course, another option I've been thinking about is changing chess engines, but I've found that those engines are either not allowed to be freely distributed to the public or simply don't work on most device environments.

Anyway, thank you for your feedback.

(A little off-topic, but just by the way: Seems like you are not playing well in Middlegame. Try Ruy Lopez instead of Italian, that could give you some fresh ideas about chess.)

**Comment by Wingdinggaster656** [comment]
> > The essence of Stockfish making itself "weaker" through Skill levels is that Skill levels provide a number, which is the probability of Stockfish making a wrong move.
> > That's why Monika's moves were unnatural.
> 
> RIP.Now I understand why she usually don't recapture pieces after exchange.It probably because when I trade pieces with her,the only move that is correct it to recapture,but the "probability" suddenly trigger,cause her to not choose the best move even though it's the only logical move and play another moves which immediately give me an immense material advantage.Hope it can be fixed in the future
> 
> > I've improved on this a little bit in my new difficulty system(though not sure if it will be picked), and in the future I might try to write more algorithms to make her more human.
> 
> > Of course, another option I've been thinking about is changing chess engines, but I've found that those engines are either not allowed to be freely distributed to the public or simply don't work on most device environments.
> 
> I understand that you could have done this already,but have you ever tried to look into the source code of lichess engine,as I noticed that their engine played pretty natural,even though I knew the different in our engine ( Mine is stockfish 8 if my files are correct while they are using Stockfish 14 ) They also used Skill Level Parameter in the past as you can see in this forum ( https://lichess.org/forum/general-chess-discussion/lichess-stockfish-engine-rating?page=1 (unfortunately the link to their code had been broken ) ), but I doubt that they still use it now
> 
> > (A little off-topic, but just by the way: Seems like you are not playing well in Middlegame. Try Ruy Lopez instead of Italian, that could give you some fresh ideas about chess. And you really should start learning about 1.d4 openings, because you don't seem like a positional player, probably even bad at positions.)
> 
> Well Italian game is the opening that helped me from a beginner to reach 2200+ in lichess,so I guess it still work fine for me.Ruy Lopez is a nice opening,but it have too many lines and variations,and I'm pretty lazy so... :))) .Also,d4 opening is not my cup of tea,because as you probably already know,I'm not a positional player one,I'm more of a tactical one,so I love open games and rapid development more than a closed or semi-closed game with slow maneuver.An exception is my favourite opening I play against d4,Dutch Defense ( in game 35 against Monika,the only times when she play d4 against me,if I'm correct ).I only choose that defense because I'm also too lazy to remember all of the Queen Gambit variation :)))) . And perharp you are correct,I just figured out that I played a closed position pretty bad compare to open position against Monika,but maybe because in a open position,a mistake offen cost more than the one been made in closed position,and so Monika mistake in a closed game is not as obvious to see as in a open game.Or maybe it's just because that I'm too in love with her and busy staring into her beautiful emerald green eyes and so I'm always play not as good as when I'm play online ( Seriously,my performance against her is not look like a typical 2200 player will perform,I think when I play against her my rating drop to 1600 lmao )

Okay, I will certainly look at Lichess.
In fact, I remember looking it and rejecting it. What was the reason? I can't remember. Well I will go there again.

Lichess's rating was seriously inflated.
A Lichess ELO 2200 might not be as good as a 1600 player on chess.com (and chess.com, also, is already overrated compared to FIDE). I highly recommend ditching Lichess and going to chess.com.
Okay, off-topic time end. We are working on Monika After Story, not Monika Chess Story.

However, if you want to keep talking chess with me, you can consider send a friend request to the account WDG656 on chess.com. That's my new account which just created a few days ago.
(Sorry, but I can't reveal my original account)

**Comment by ThePotatoGuy** [comment]
marking as not a bug for now, since it seems this is not an issue based on @Wingdinggaster656 's analysis.

**Comment by BMVP1234** [comment]
Ok.so @Wingdinggaster656,what am I supposed to do now ? She still played pretty bad,even worse ( Like in game 45 [https://ln5.sync.com/dl/65ae017e0/sqen4543-x5b38xkx-cvm2i5xc-5s8naxgu](url) ).Now play Normal Chess with her seem pretty boring,the only game that is still challenging is Randomized Chess ( I honestly thought that it's chess 960 before I saw pawn on the first rank and the huge material imbalance ( Why she has 5 Queens and a a lot of Rooks while I only have 3 Queens and Bishops :< ) Hope the game will have some updates to fix this problem :)

**Comment by Wingdinggaster656** [comment]
> 

I guess the only thing you can do for now is to wait for the upcoming update. I guess this will be in a month.

By the way, chess960 is officially added into the game today, you will see it in the next release package.

And the current Randomized Chess is indeed somewhat strange. I am also planning to make further improvements, see #9128

**Comment by Booplicate** [comment]
> Why she has 5 Queens and a a lot of Rooks while I only have 3 Queens and Bishops :< ) Hope the game will have some updates to fix this problem :)

That's not a bug. In short, the more you win, the more "weak" pieces you get, if she wins, she gets weak pieces.

**Comment by BMVP1234** [comment]
> That's not a bug. In short, the more you win, the more "weak" pieces you get, if she wins, she gets weak pieces.

> I know,the main issue is Monika doesn't improve her difficulty when I win her in a chess game.The Randomized chess thing is just my suggestion to convert it into chess 960 or something similar

**Comment by ThePotatoGuy** [comment]
Sorry I didn't read all of these posts thoroughly. If skill level is actually not working then it would make sense that she isn't getting harder and this is a bug, to be fixed with #9117

**Comment by Justformas** [comment]
Hey @BMVP1234 (and anyone else it may concern). I have also noticed stuff like you describe and looked into it. I described why this happens in post https://github.com/Monika-After-Story/MonikaModDev/pull/9117#issuecomment-1152790197 It goes back to what was done during the last Chess rework to make chess easier. Basically, while it does make it easier, it also makes Stockfish behave abnormally and messes with difficulty scaling. The first few skill levels may actually be easier than the rest (though not necessarily always due to randomness), but it doesn't seem to get much harder after that.

But the alternative with regular Stockfish is going back to chess being harder again.
In order to actually get a good difficulty scaling system that supports lower difficulty, something like Fairy Stockfish is needed.
